agent/grpc/client.go

```diff
@@ -102,22 +157,39 @@ func (c *ClientConnPool) dial(datacenter string, serverType string) (*grpc.Clien
 
 // newDialer returns a gRPC dialer function that conditionally wraps the connection
 // with TLS based on the Server.useTLS value.
-func newDialer(servers ServerLocator, wrapper TLSWrapper, useTLSForDC func(dc string) bool) func(context.Context, string) (net.Conn, error) {
-	return func(ctx context.Context, addr string) (net.Conn, error) {
-		d := net.Dialer{}
-		conn, err := d.DialContext(ctx, "tcp", addr)
+func newDialer(cfg ClientConnPoolConfig, gwResolverDep *gatewayResolverDep) func(context.Context, string) (net.Conn, error) {
+	return func(ctx context.Context, globalAddr string) (net.Conn, error) {
+		server, err := cfg.Servers.ServerForGlobalAddr(globalAddr)
 		if err != nil {
 			return nil, err
 		}
 
-		server, err := servers.ServerForAddr(addr)
+		if cfg.DialingFromServer &&
+			gwResolverDep.GatewayResolver != nil &&
+			cfg.ALPNWrapper != nil &&
+			server.Datacenter != cfg.DialingFromDatacenter {
+			// NOTE: TLS is required on this branch.
+			conn, _, err := pool.DialRPCViaMeshGateway(
+				ctx,
+				server.Datacenter,
+				server.ShortName,
+				cfg.SrcAddr,
+				tlsutil.ALPNWrapper(cfg.ALPNWrapper),
+				pool.ALPN_RPCGRPC,
+				cfg.DialingFromServer,
+				gwResolverDep.GatewayResolver,
+			)
+			return conn, err
+		}
+
+		d := net.Dialer{LocalAddr: cfg.SrcAddr, Timeout: pool.DefaultDialTimeout}
```

Comment:
Made the non-gateway version of grpc also have a non-infinite dial timeout like all of the others.